File: leetcode/0003-longest-substring-without-repeating-characters/solution.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int maxLen = 0;
        int[] arr = new int[256];
        int l = 0, r = 0;

        while(r < s.length()){
            char c = s.charAt(r);
            l = (arr[c] > 0) ? Math.max(arr[c], l) : l;
            arr[c] = r + 1;
            maxLen = Math.max(maxLen, r - l + 1);
            r++;
        }
        
        return maxLen;
    }
}

File: leetcode/0003-longest-substring-without-repeating-characters/solution.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        i = 0
        n = len(s)
        ans = 0
        seen = set()

        for j in range(n):
            while s[j] in seen:
                # keep removing from left until we don't eliminate the element our right is standing on
                seen.remove(s[i])
                i += 1
            
            seen.add(s[j])
            ans = max(ans, j - i + 1)

        return ans
